### PR TITLE
More XML

### PR DESCRIPTION
**Pull request by davemevans**
- MID is serialised to multiple SegmentationUpid Nodes
- upid length is handled more sensibly for some cases (but still not great)
- optional attributes handled gracefully if missing in various places

### CODE REVIEW

**Comment by davemevans** [comment]
Making changes .. don't merge just yet.

**Comment by futzu** [comment]
Good call, I was looking up the MID stuff last night.  Remember, extra attributes are allowed. If something isn't clear, make up an attribute if you feel the need. I add a lot of stuff to threefive.  If a parser doesn't know it, they ignore it. 

I you print anything, use threefive.print2, it writes to stderr, so we don't breaking piping | .

If you're feeling super extra crazy, you could fix Eidr upids I am using the Eidr compact binary format, 
and it's supposed to be the bizarre and bloated and convoluted  text format  for Dash .  

I am not a fan of EIDR.  :)

**Comment by davemevans** [comment]
EIDR added, and I actually made MID() work on the third attempt. A few other bits in there that would be worth a review if you got time - XML escaping and some other non-XML specific fixes.
Have a good weekend!

**Comment by davemevans** [comment]
Please feel free to reject or refactor any of this stuff as you see fit - I am not precious about it.

**Comment by davemevans** [comment]
Plan is to have a look at refactoring the UPID stuff out to UPIDs and try to complete the xml() bits to a reasonable standard then see where we are.

I'm a bit confused about how the Upids work since they store some of the context but not all of it and other bits are stored in the descriptor, but I'll have a think. Won't be quick most likely.

File: threefive/descriptors.py [hunk 1]
```diff
@@ -484,21 +479,29 @@ def xml(self):
...
-            })
-
-        for upid in upids:
-            sd.add_child(Node('SegmentationUpid',attrs={
-                'segmentation_upid_type': upid["upid_type"],
-            }, value=upid["segmentation_upid"]))
-
+            ud_attrs={'segmentation_upid_type': self.segmentation_upid_type,}
+            value = self.segmentation_upid
+            if self.segmentation_upid_type == 12:
```
Comment (futzu) [inline threefive/descriptors.py]: hex for upid types

File: threefive/descriptors.py [hunk 2]
```diff
@@ -484,21 +479,29 @@ def xml(self):
...
-                "upid_length": self.segmentation_upid_length,
-                "segmentation_upid": self.segmentation_upid,
-            })
-
-        for upid in upids:
-            sd.add_child(Node('SegmentationUpid',attrs={
-                'segmentation_upid_type': upid["upid_type"],
-            }, value=upid["segmentation_upid"]))
-
+            ud_attrs={'segmentation_upid_type': self.segmentation_upid_type,}
```
Comment (futzu) [inline threefive/descriptors.py]: None of this upid logic should be in descriptors.

File: threefive/stuff.py [hunk 3]
```diff
@@ -11,3 +11,9 @@ def print2(stuff=b""):
     """
     print(stuff, file=stderr, flush=True)
 
+
+class dottable_dict(dict):
```
Comment (futzu) [inline threefive/stuff.py:15]: Why?

File: threefive/descriptors.py [hunk 4]
```diff
@@ -488,7 +483,22 @@ def xml(self):
                 'archive_allowed_flag': self.archive_allowed_flag,
                 'device_restrictions': k_by_v(table20, self.device_restrictions),
             }))
-        sd.add_child(ud)
+
+        upids = []
+        if self.segmentation_upid_type == 13:
```
Comment (futzu) [inline threefive/descriptors.py:515]: This should not be in descriptors. 
We identify UPIDS in hex, always match the format used in the map, 
there is a command_map, descriptor_map and upid_map
command and descriptor are in base10, for some reason they put the upids in base166

```py3
 if self.segmentation_upid_type == 0x0d:   
```

Comment (futzu) [reply threefive/descriptors.py:515]: I use maps a lot. They help to cut down on conditionals and keep the code clean

if you look at the upid class, the parent class for the upids, is really an interface ,
every subclass implements decode() and encode() the same way. 

 ```py3

class Upid:
    """
    Upid base class handles URI UPIDS
    """

    def decode(self):
        """
        decode Upid
        """
 
    def encode(self, nbin, seg_upid):
        """
        encode Upid
        """
    

``` 

The reason for this is so you can do 
```py3
        the_upid = upid_map[upid_type][1](
            None, upid_type, self.segmentation_upid_length
        )
        the_upid.encode(nbin, self.segmentation_upid)
```

the_upid in the code can be any upid, we don't have be concerned with which one. 

the commands and descriptors are done the sane way

If you notice all the  __from_xml__ and __xml__ methods are all called the same way.

This saves a bunch of time, and a bunch of code.

Comment (futzu) [reply threefive/descriptors.py:515]: if upids have a __from_xml__  method  it  could be called the same way, and without conditionals

```py3
  the_upid = upid_map[upid_type][1](
            None, upid_type, self.segmentation_upid_length
        )
 the_pid.from_xml(self,stuff)
```

That puts the logic with the object, that is always a good thing.

**Review by futzu — COMMENTED** [review]
Here's some feedback, I should have explained things better. I apologize.

File: threefive/descriptors.py (hunk 4 above)
Comment (futzu) [reply threefive/descriptors.py:515]: Cyclomatic Complexity is super important, although it's directly a metric for quality, low cyclomatic complexity itend to lead to the best code. Install radon via pip.
`radon cc -sa threefive/`
```js
.....
298 blocks (classes, functions, methods) analyzed.
Average complexity: A (2.1040268456375837)
```
A lot people say a score of ten is okay, I think ten  is way to high, my target is between  two and three.
If a method scores over 3, split it .

**Comment by davemevans** [comment]
Great feedback, thanks. I'll take another look.

**Pull request closed without merging**